### PR TITLE
Granda Mento: support for approving, cancelling, and executing exchange proposals

### LINKED ISSUE #7945: Allow GrandaMento to mint stable tokens
Currently, `StableToken.sol` only allows its Exchange or the `Validators` contract to call `mint` according to this [`require`](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L225)

As mentioned in the [Granda Mento CIP](https://github.com/celo-org/celo-proposals/pull/224/files), GrandaMento must be able to call `mint`.

`mint` should only be callable by:
* the address for `getExchangeRegistryId()` found in the Registry
* the address for Validators found in the Registry
* the address for GrandaMento found in the Registry

This should be entirely backward compatible.

### LINKED ISSUE #7946: Allow Granda Mento to burn stable tokens
Currently, `StableToken.sol` only allows its Exchange to call `burn` according to this [modifier call](https://github.com/celo-org/celo-monorepo/blob/master/packages/protocol/contracts/stability/StableToken.sol#L275)

As mentioned in the [Granda Mento CIP](https://github.com/celo-org/celo-proposals/pull/224/files), GrandaMento must be able to call `burn`.

`burn` should only be callable by:
* the address for getExchangeRegistryId() found in the Registry
* the address for GrandaMento found in the Registry

This should be entirely backward compatible.

### LINKED ISSUE #7954: GrandaMento.sol - implement exchange execution
Implement exchange execution according to the [GrandaMento CIP](https://github.com/celo-org/celo-proposals/pull/224).

- [x] Add the configurable state variable `exchangeWaitPeriodSeconds`.
- [x] Create a function that lets the owner of GrandaMento set `exchangeWaitPeriodSeconds`. Should emit an event.
- [x] Implement `executeExchange`. Should mint/burn the StableToken as appropriate, and pay to/spend the reserve, while imposing a spread on the exchange. Should emit an event.
- [x] Write extensive tests!

Expected end state:
An approved exchange that's passed the waiting period can be executed in its entirety.

### LINKED ISSUE #7982: GrandaMento.sol - add approver and implement exchange proposal approval
Implement the ability for the designated approver to approve an exchange proposal as defined by the [GrandaMento CIP](https://github.com/celo-org/celo-proposals/pull/224)

- [x] add the approver state variable
- [x] add function for the owner of GrandaMento to set the approver. Should emit an event.
- [x] implement `approveExchangeProposal`. Should emit an event.
- [x] add extensive tests!

Expected end state:

The approver can be set by the owner, and the approver can approve exchange proposals.

### LINKED ISSUE #7983: GrandaMento.sol - implement exchange proposal cancellation
Implement the ability for an exchange proposal to be cancelled as defined by the [GrandaMento CIP](https://github.com/celo-org/celo-proposals/pull/224)

- [x] Implement `cancelExchangeProposal`. Should emit an event.
- [x] Write extensive tests!

Expected end state:
When an exchange proposal is in the appropriate state defined by the CIP, exchanges are cancelled and the deposit is refunded to the proposer. This is both capturing states that are cancelled by the proposer or governance. The demurrage enabled case should be captured here when refunding funds.

### PR DESCRIPTION
**Pull request by tkporter**
### Description

This PR brings Granda Mento to being fully functioning! Some highlights:
* Adds the `approver` and a corresponding setter
* Adds `approveExchangeProposal`
* Adds `cancelExchangeProposal`
* Adds the `vetoPeriodSeconds` and a corresponding setter
* Adds `executeExchangeProposal`
* Renamed the `ExchangeState` enum to `ExchangeProposalState`
* Stable tokens are now specified by their Registry ID when creating a new exchange proposal and for keeping track of exchange limits
* Adds `getActiveProposalIds`, which is a view function to see all proposals that are in the `Proposed` or `Approved` state. They're kept track of in a linkedlist for O(1) inserts/removals rather than an array. If it were kept track of in an array, cancelling or executing a proposal could involve the sender specifying the index of the proposal in the array resulting in O(1) removals, but because there is a long delay for governance to cancel a proposal, it's possible that the index specified in the veto governance proposal would be incorrect by the time the proposal is actually executed.
* `StableToken.sol` allows GrandaMento to mint and burn stable tokens
* Introduces `assertRevertWithReason`, which ensures that a transaction or call reverts with a specific reason

### Other changes

* Made a clarifying comment in Reserve.sol

### Tested

Unit tests :stuck_out_tongue:

### Related issues

- Fixes #7982
- Fixes #7983
- Fixes #7954
- Fixes #7945
- Fixes #7946

### Backwards compatibility

Fully backward compatible, does bump the minor version of StableToken and StableTokenEUR however

### Documentation

N/a

### CODE REVIEW

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 1]
```diff
@@ -54,7 +70,7 @@ contract GrandaMento is
 
   struct ExchangeProposal {
     // The exchanger/proposer of the exchange proposal.
-    address exchanger;
+    address payable exchanger;
```
Comment (tkporter) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Considering changing this to `proposer`, any thoughts?

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 2]
```diff
@@ -33,17 +33,32 @@ contract GrandaMento is
...
   // Emitted when the spread is set.
   event SpreadSet(uint256 spread);
 
+  // Emitted when the veto period in seconds is set.
+  event VetoPeriodSecondsSet(uint256 vetoPeriodSeconds);
+
   // Emitted when the exchange limits for a stable token are set.
   event StableTokenExchangeLimitsSet(
-    address indexed stableToken,
+    string stableTokenRegistryId,
```
Comment (tkporter) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Not indexed because an indexed string will actually emit the value equivalent to `keccak(abi.encodePacked(strValue))` rather than `strValue`. I figured that indexed events aren't *that* useful because presumably a chain indexing service will still be able to work regardless of the parameter being indexed or not.

An alternative is to emit both:
`string stableTokenRegistryId, bytes32 indexed stableTokenRegistryIdHash`

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 3]
```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
...
+      address stableToken = proposal.stableToken;
+      sellToken = IERC20(stableToken);
+      // When selling stableToken, the sell amount is stored in units.
+      // Units must be converted to value when refunding.
+      sellAmount = IStableToken(stableToken).unitsToValue(proposal.sellAmount);
+    }
+    // In the event a precision issue from the unit <-> value calculations results
+    // in sellAmount being greater than this contract's balance, set the sellAmount
+    // to the entire balance.
+    // This check should not be necessary for CELO, but is done so regardless
```
Comment (tkporter) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Curious to hear thoughts here

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 1 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: exchanger is consistent from what I have seen in other contrats I think

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 4]
```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
...
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require that the proposal is in the Approved state.
+    require(proposal.state == ExchangeProposalState.Approved, "Proposal must be in Approved state");
+    // Require that the veto period has elapsed since the approval time.
+    require(
+      proposal.approvalTimestamp.add(vetoPeriodSeconds) <= block.timestamp,
+      "Veto period not elapsed"
+    );
+    // Mark the proposal as executed.
+    proposal.state = ExchangeProposalState.Executed;
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: nit: to make debugging more straightforward I'd place this right before the `emit`

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 5]
```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
...
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require the appropriate state and sender.
+    // This will also revert if a proposalId is given that does not correspond
+    // to a previously created exchange proposal.
+    require(
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Is it easier to debug if we separate this require with a different text for the different cases?

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 6]
```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
...
+      (proposal.state == ExchangeProposalState.Proposed && proposal.exchanger == msg.sender) ||
+        (proposal.state == ExchangeProposalState.Approved && isOwner()),
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Can't the owner cancel on any stage?

File: packages/protocol/contracts/stability/StableToken.sol [hunk 7]
```diff
@@ -41,6 +41,8 @@ contract StableToken is
 
   event TransferComment(string comment);
 
+  bytes32 constant GRANDA_MENTO_REGISTRY_ID = keccak256(abi.encodePacked("GrandaMento"));
```
Comment (martinvol) [inline packages/protocol/contracts/stability/StableToken.sol:44]: Is it possible to do `registry.getAddressForStringOrDie` here?

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 3 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Trying to think of an edge case but I can't think of one. To be cautious I'd try to avoid the check for CELO, but not really a strong thought.

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 6 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Hmm the CIP originally specified the owner can only cancel to "veto" an approval (ie when the proposal is in the Approved state). Could you think of a situation where it makes sense to have a governance proposal cancel an exchange proposal when it hasn't been approved?

File: packages/protocol/contracts/stability/StableToken.sol (hunk 7 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/StableToken.sol:44]: View calls like that aren't allowed outside of the constructor or functions. Also because the address may change in the registry it's best to use the registry ID

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 5 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Imo it's easily implied by the sender. I like the setup of having a single require statement rather than if statements depending on the state/sender, wdyt?

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 6 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: yeah makes sense

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 8]
```diff
@@ -152,32 +197,167 @@ contract GrandaMento is
...
+    emit ExchangeProposalApproved(proposalId);
+  }
+
+  /**
+   * @notice Cancels an exchange proposal.
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: why is this `nonReentrant` if it's only calling core contracts in this function?

Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: The pattern I've seen so far throughout our contracts is if a function is state-modifying and involves any calls to an outside contract, we use `nonReentrant` just to be super safe. Apart from the gas cost involved with the `nonReentrant` modifier I don't see a big downside in having it, wdyt?

**Comment by ghost** [comment]
Would it make sense to add a new folder packages/protocol/contracts/liquidity and put GrandaMento.sol in there? 
It feels weird that GrandaMento.sol is in the stability folder as it is liquidity without a stability component. In longer-term stability and liquidity would probably be more separated anyways? @tkporter @martinvol

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 4 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: I changed this but thought about it some more and I think updating the state before doing any token transfers/minting/burning is the right move. Doing so is a common practice against any kind of reentrancy, which is ofc prevented by the `nonReentrant` modifier, but just to err on the side of safety may be good to keep

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 9]
```diff
@@ -152,32 +197,167 @@ contract GrandaMento is
...
       sellToken.transferFrom(msg.sender, address(this), sellAmount),
-      "Transfer of sell token failed"
+      "Transfer in of sell token failed"
     );
 
     // Record the proposal.
-    uint256 proposalId = exchangeProposalCount;
-    exchangeProposals[proposalId] = ExchangeProposal({
+    // Add 1 to the running proposal count and use as the proposalId. This is
+    // intentionally never 0 because LinkedList, which is used for keeping track
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Technically, it's 0 when there hasn't been any exchange proposal created, right?

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 8 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: yeah, only worried about gas usage, not a big deal actually.

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 5 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: oh you're right, the `||` makes it require more than two requires, I'd just leave like it is 👍

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 10]
```diff
@@ -152,32 +197,167 @@ contract GrandaMento is
...
+    // to a previously created exchange proposal.
+    require(
+      (proposal.state == ExchangeProposalState.Proposed && proposal.exchanger == msg.sender) ||
+        (proposal.state == ExchangeProposalState.Approved && isOwner()),
+      "Sender cannot cancel the exchange proposal"
+    );
+    // Mark the proposal as cancelled. Do so prior to refunding as a measure against reentrancy.
+    proposal.state = ExchangeProposalState.Cancelled;
+    // Remove the proposal from the active list.
+    activeProposalIds.remove(proposalId);
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: just double checking, IIUC this list has an index of elements, so this call is `O(1)` instead of a `O(N)` lookup, right?

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 11]
```diff
@@ -152,32 +197,167 @@ contract GrandaMento is
...
+      );
+    } else {
+      // If the exchange is selling stable token, the stable token is burned from
+      // this contract and CELO is transferred from the Reserve to the exchanger.
+
+      // Burn the stable token from this contract.
+      require(IStableToken(proposal.stableToken).burn(sellAmount), "Stable token burn failed");
+      // Transfer the CELO from the Reserve to the exchanger.
+      require(
+        getReserve().transferExchangeGold(proposal.exchanger, proposal.buyAmount),
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: The idea is to have Granda Mento in CR5, leaving it unusable as all the parameters are zero, and then make a follow up proposal for activation setting the limits, setting the approver and a `Reserve.addExchangeSpender` call?

If so ,for me it was useful to keep a checklist for the proposals and even write a draft json (although we still don't have a clear place to check those in the repo, there are some jsons in the celo-proposals repo)

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 12]
```diff
@@ -222,46 +440,37 @@ contract GrandaMento is
     emit SpreadSet(newSpread);
   }
 
+  /**
+   * @notice Sets the veto period in seconds.
+   * @dev Sender must be owner.
+   * @param newVetoPeriodSeconds The new value for the veto period in seconds.
+   */
+  function setVetoPeriodSeconds(uint256 newVetoPeriodSeconds) public onlyOwner {
+    vetoPeriodSeconds = newVetoPeriodSeconds;
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: shall we safety check for a minimum? It'd prevent an (unlikely) fat finger.

**Comment by martinvol** [comment]
@tobikuhlmann  Makes sense to move to a new folder to me!

File: packages/protocol/contracts/stability/GrandaMento.sol [hunk 13]
```diff
@@ -212,6 +392,44 @@ contract GrandaMento is
...
+    require(rateDenominator > 0, "No oracle rates present for token");
+    return FixidityLib.wrap(rateNumerator).divide(FixidityLib.wrap(rateDenominator));
+  }
+
+  /**
+   * @notice Sets the approver.
+   * @dev Sender must be owner.
+   * @param newApprover The new value for the spread.
+   */
+  function setApprover(address newApprover) public onlyOwner {
```
Comment (martinvol) [inline packages/protocol/contracts/stability/GrandaMento.sol]: Audit likely will flag approver can be the zero address, and I think that is on propose, so I'll document it in the natspec.

Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: (and add a test, wil cEUR we realized we couldn't set something as zero after the audit)

File: packages/protocol/test/stability/grandamento.ts [hunk 14]
```diff
@@ -396,23 +456,556 @@ contract('GrandaMento', (accounts: string[]) => {
...
+            const reserveBalanceBefore = await goldToken.balanceOf(reserve.address)
+            await grandaMento.executeExchangeProposal(1)
+            const grandaMentoBalanceAfter = await goldToken.balanceOf(grandaMento.address)
+            const reserveBalanceAfter = await goldToken.balanceOf(reserve.address)
+
+            assertEqualBN(reserveBalanceAfter.minus(reserveBalanceBefore), celoSellAmount)
+            assertEqualBN(grandaMentoBalanceBefore.minus(grandaMentoBalanceAfter), celoSellAmount)
+          })
+
+          it('transfers the entire CELO balance to the Reserve if the sell amount is higher than the balance', async () => {
```
Comment (martinvol) [inline packages/protocol/test/stability/grandamento.ts]: In what scenario this would be hit?

File: packages/protocol/test/stability/grandamento.ts [hunk 15]
```diff
@@ -396,23 +456,556 @@ contract('GrandaMento', (accounts: string[]) => {
...
+        // proposalId 1
+        await grandaMento.createExchangeProposal(
+          stableTokenRegistryId,
+          stableTokenSellAmount,
+          false,
+          {
+            from: alice,
+          }
+        )
+        // proposalId 1
```
Comment (martinvol) [inline packages/protocol/test/stability/grandamento.ts]:
```suggestion
        // proposalId 2
```

File: packages/protocol/test/stability/grandamento.ts [hunk 16]
```diff
@@ -396,23 +456,556 @@ contract('GrandaMento', (accounts: string[]) => {
...
+        await grandaMento.createExchangeProposal(
+          stableTokenRegistryId,
+          stableTokenSellAmount,
+          false,
+          {
+            from: alice,
+          }
+        )
+        // proposalId 2
```
Comment (martinvol) [inline packages/protocol/test/stability/grandamento.ts]:
```suggestion
        // proposalId 3
```

File: packages/protocol/test/stability/grandamento.ts [hunk 17]
```diff
@@ -396,23 +456,556 @@ contract('GrandaMento', (accounts: string[]) => {
...
+          const aliceBalanceAfter = await stableToken.balanceOf(alice)
+
+          assertEqualBN(
+            grandaMentoBalanceBefore.minus(grandaMentoBalanceAfter),
+            stableTokenSellAmount
+          )
+          assertEqualBN(aliceBalanceAfter.minus(aliceBalanceBefore), stableTokenSellAmount)
+        })
+
+        it('refunds the appropriate stable token amount value when the inflation factor is 1', async () => {
```
Comment (martinvol) [inline packages/protocol/test/stability/grandamento.ts]:
```suggestion
        it('refunds the appropriate stable token amount value when the inflation factor is above 1', async () => {
```

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 3 above)
Comment (martinvol) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Thinking more about this, what would happen in the case there are many concurrent exchange proposals and we have had inflation with cUSD?

IIUC, then the ones that that are executed first will get the correct amount of stabletoken and the last one will get the remaining, effectively benefiting the ones executed first. It this an issue or just a rounding error?

File: packages/protocol/test/stability/grandamento.ts [hunk 18]
```diff
@@ -396,23 +456,556 @@ contract('GrandaMento', (accounts: string[]) => {
...
+      )
+    })
+  })
+
+  describe('#executeExchangeProposal', () => {
+    const stableTokenSellAmount = unit.times(500)
+    const celoSellAmount = unit.times(100)
+    let sellCelo = false
+    beforeEach(async () => {
+      if (sellCelo) {
```
Comment (martinvol) [inline packages/protocol/test/stability/grandamento.ts]: Can this if be removed?

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 9 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Just changed the wording of the comment, it was confusing. What I was trying to say is that proposal IDs start at 1 rather than 0 because of the limitation imposed by LinkedList that the key must be non-zero, and that's why `exchangeProposalCount` is incremented first

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 10 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Correct, I added a comment as well indicating it's O(1)

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 11 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: So I'm thinking if we have made up our mind already for the parameters that are set in the `initialize` function (approver, spread, and the veto period), then we can set those via CR5, and then there would need to be a follow-up governance proposal to call `addExchangeSpender` and set the exchange limits for cUSD and cEUR. https://github.com/celo-org/celo-proposals/pull/231 is the source of truth for that right now-- haven't had the time to craft the jsons yet though

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 3 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: Yeah if there were rounding errors in the unit <-> value conversion, the last one would get slightly less stabletoken in the end. But this should only be a handful of wei at most because of the rounding errors, so imo it's negligible and hard to avoid

File: packages/protocol/contracts/stability/GrandaMento.sol (hunk 12 above)
Comment (tkporter) [reply packages/protocol/contracts/stability/GrandaMento.sol]: I was thinking it'd be neat to allow governance to decide they want to be the approver themselves rather than the multisig down the road. Meaning governance would have the ability to:
1. set Governance as the approver
2. set `vetoPeriodSeconds` to be 0 so the proposal can be executed immediately after a governance approval updates it

Thoughts? Tbh it's not a super likely scenario. But I'm not super worried about fat fingers because `setVetoPeriodSeconds` will need to be called via a governance proposal, which will receive a lot of scrutiny

File: packages/protocol/test/stability/grandamento.ts (hunk 18 above)
Comment (tkporter) [reply packages/protocol/test/stability/grandamento.ts]: Just added a comment explaining why it's needed for sellCelo and not !sellCelo. I figure it's best to not run it unless needed to keep tests more concise in what state they modify

File: packages/protocol/test/stability/grandamento.ts (hunk 14 above)
Comment (tkporter) [reply packages/protocol/test/stability/grandamento.ts]: I don't see a situation tbh, that's why I'm on the fence about https://github.com/celo-org/celo-monorepo/pull/8082/files#diff-53b0c8ae012f6e9215f6c0216a5c6bbf74d4abce2fd1c1cd636e502d76be7608R353. It's a pretty common practice to have that sort of safety just in case to ensure that the call will never revert from an unforeseen error, though. I'm curious to hear Yorke/CAP's opinion
